detect_regressions.py

```diff
@@ -0,0 +1,184 @@
+                        reg_df.loc[f"{(runtime, name)}"] = [
+                            category,
+                            "duration",
+                            stats["duration_mean"],
+                            stats["duration_last"],
+                            stats["duration_last-1"],
+                            stats["duration_last-2"],
+                            dur_threshold,
+                        ]
+
+                    if (
+                        stats["avg_memory_last"] >= avg_mem_threshold
+                        and stats["avg_memory_last-1"] >= avg_mem_threshold
+                        and stats["avg_memory_last-2"] >= avg_mem_threshold
+                    ):
+                        reg = (
+                            f"{runtime= }, {name= }, {category= }, "
+                            f"avg_mem_last = "
+                            f"{(stats['avg_memory_last'], stats['avg_memory_last-1'], stats['avg_memory_last-2'])}, "
+                            f"{avg_mem_threshold= } \n"
+                        )
+
+                        regressions.append(reg)
+
+                        reg_df.loc[f"{(runtime, name)}"] = [
+                            category,
+                            "avg_memory",
+                            stats["avg_memory_mean"],
+                            stats["avg_memory_last"],
+                            stats["avg_memory_last-1"],
+                            stats["avg_memory_last-2"],
+                            avg_mem_threshold,
+                        ]
+
+                    if (
+                        stats["peak_memory_last"] >= peak_mem_threshold
+                        and stats["peak_memory_last-1"] >= peak_mem_threshold
+                        and stats["peak_memory_last-2"] >= peak_mem_threshold
+                    ):
+                        reg = (
+                            f"{runtime= }, {name= }, {category= }, "
+                            f"peak_mem_last = "
+                            f"{(stats['peak_memory_last'], stats['peak_memory_last-1'], stats['peak_memory_last-2'])}, "
+                            f"{peak_mem_threshold= } \n"
+                        )
+
+                        regressions.append(reg)
+
+                        reg_df.loc[f"{(runtime, name)}"] = [
+                            category,
+                            "peak_memory",
+                            stats["peak_memory_mean"],
+                            stats["peak_memory_last"],
+                            stats["peak_memory_last-1"],
+                            stats["peak_memory_last-2"],
+                            peak_mem_threshold,
+                        ]
+            else:
+                # df_pass is empty or we don't have enough stats
+                pass
+
+    return reg_df, regressions
+
+
+def regressions_report(reg_df, regressions):
+
+    # write reg_df to markdown for GHA summary
+    reg_df.to_markdown("regressions_summary.md")
```

Comment:
It should also be possible to write directly to `$GITHUB_STEP_SUMMARY` here and avoid the intermediate file, though it's fine as is.

Reply:
I'm not sure I follow how we can write directly to the summary without doing a `to_markdown` step.

Reply:
I'm not proposing skipping the `to_markdown` step, just that we could probably directly write to `$GITHUB_STEP_SUMMARY` here. Something like

```python
with open(os.environ.get("GITHUB_STEP_SUMMARY", "regressions_summary.md") mode="a") as f:
    reg_df.to_markdown(f)
```
then drop the additional step.

Doesn't really matter too much, though

Reply:
Oh I see, yes I think I like it in the action now because it makes it more clear what's happening, at least to me. But we could re-evaluate in the future